Set up continuous integration

loicgilbert opened the pull request:
Built project on Linux and Windows, with python 3.6 and 3.7 on both using Appveyor.
(close #10, close#11 )

loicgilbert committed:
add appveyor.yml

loicgilbert committed:
test CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-Ci

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-Ci

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-Ci

loicgilbert committed:
test-Ci

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

loicgilbert committed:
test-CI

Test if build fail when error in program on windows

loicgilbert committed:
test if build is working on windows and linux

loicgilbert committed:
forgot to uncoment lines

loicgilbert committed:
test if build don't work when error in program

loicgilbert committed:
build artifact on windows

loicgilbert committed:
fix artifact building

loicgilbert committed:
build artifacts on windows and linux

loicgilbert committed:
Fix artifact building o windows and linux

loicgilbert committed:
Push executable as artifact

loicgilbert committed:
fix artifact creation

loicgilbert committed:
Fix artifacts creation

loicgilbert committed:
Fix pyinstaller on linux

loicgilbert committed:
Fix pyinstaller on linux

loicgilbert committed:
Fix pyinstaller on linux

loicgilbert committed:
Fix pyinsaller on linux

loicgilbert committed:
Fix small mistakes

loicgilbert committed:
Implement CI

loicgilbert committed:
Implement CI

loicgilbert committed:
Delete null

loicgilbert committed:
Check appveyor build

pavel-kirienko commented on appveyor.yml:
```diff
@@ -0,0 +1,58 @@
+  - cmd: "python -m pip install mypy"
+  - cmd: "python -m  pip install -r requirements.txt"
+  - sh: "pip install mypy"
+  - sh: "pip install -r requirements.txt"
```
There should be no need to install the build dependencies individually; I suggest having them listed in `requirements.txt`. It might make sense to move development-only dependencies into `requirements-dev.txt` though, as is usually done.

loicgilbert replied:
Done

pavel-kirienko commented on appveyor.yml:
```diff
@@ -0,0 +1,58 @@
+
+test_script:
+  - cmd: "test_windows.bat"
+  - sh: "export QT_QPA_PLATFORM=offscreen"
```
Perhaps we should link the appropriate doc page that describes the setting `QT_QPA_PLATFORM=offscreen`? A quick google search did not return much useful info.

loicgilbert replied:
Found that, but not very useful:
https://doc.qt.io/qt-5/embedded-linux.html
Without it, there is an error during linux builds related to QT plugins.
That solution was found here:
https://github.com/ariya/phantomjs/issues/14376
I thinks it disable the software display, which is not available on appveyor.

pavel-kirienko replied:
Cool; please add these links as comments to the CI config file.

pavel-kirienko commented on appveyor.yml:
```diff
@@ -0,0 +1,58 @@
+
+after_test:
+  - cmd: "mypy zubax-kucher || ver > nul"
+  - sh: "(mypy zubax-kucher; exit 0)"
```
Why explicit exit? Why not just `mypy zubax-kucher`?

pavel-kirienko commented on appveyor.yml:
```diff
@@ -0,0 +1,58 @@
+  - sh: "(mypy zubax-kucher; exit 0)"
+  - cmd: "pep8 --max-line-length=120 --exclude libraries || ver > nul"
+  - sh: "(pep8 --max-line-length=120 --exclude libraries; exit 0)"
+  - cmd: "7z a zubax-kucher.zip *"
```
As I understood, we create a 7z archive, not a zip, is that correct? If yes, shouldn't the extension be `.7z`?

pavel-kirienko commented on appveyor.yml:
```diff
@@ -0,0 +1,58 @@
+  - cmd: "mypy zubax-kucher || ver > nul"
+  - sh: "(mypy zubax-kucher; exit 0)"
+  - cmd: "pep8 --max-line-length=120 --exclude libraries || ver > nul"
+  - sh: "(pep8 --max-line-length=120 --exclude libraries; exit 0)"
```
Please move the settings into `setup.cfg`. Example: https://github.com/UAVCAN/pyuavcan/blob/uavcan-v1.0/setup.cfg.

It is probably best to move the whole testing pipeline into the test scripts you provided to make local testing easier, would you agree?

pavel-kirienko commented on appveyor.yml:
```diff
@@ -0,0 +1,58 @@
+  - cmd: "7z a zubax-kucher.zip *"
+  - sh: "7z a zubax-kucher.zip * &>/dev/null &"
+  - "appveyor PushArtifact zubax-kucher.zip"
+  - "pyinstaller pyinstaller.spec"
```
Do I get it right that the original spec file did not require any modifications to produce a Windows executable?

loicgilbert replied:
Yep, it also works with windows exe

pavel-kirienko commented on kucher/__init__.py:
```diff
@@ -33,9 +33,17 @@
     os.path.join(THIRDPARTY_PATH_ROOT, 'quamash'),
 ]
 
+ #
```
```suggestion
```

pavel-kirienko commented on kucher/__init__.py:
```diff
@@ -33,9 +33,17 @@
+ # module located in the 'libraries' directory will be used. It is not compatible with Python 3.7, so we only declare
+ # its path if Python version is below 3.7. Otherwise, the built-in module will be used by default.
+ #
+ 
```
```suggestion
```

pavel-kirienko commented on kucher/__init__.py:
```diff
@@ -33,9 +33,17 @@
+ # 'dataclasses' module is included in Python libraries since version 3.7. For Python versions below, the dataclass
+ # module located in the 'libraries' directory will be used. It is not compatible with Python 3.7, so we only declare
+ # its path if Python version is below 3.7. Otherwise, the built-in module will be used by default.
+ #
```
```suggestion
```

pavel-kirienko commented on requirements.txt:
```diff
@@ -25,5 +25,4 @@ pytest >= 3.3
 
 # Packaging tools
 PyInstaller ~= 3.3
-patchelf-wrapper~=1.0
```
Can't use staticx without patchelf-wrapper. Could you remind me please why did it have to go?

loicgilbert replied:
patchelf-wrapper is not compatible with Windows. I removed it from requirements-dev-windows but left it on requirements-dev-linux. It is used only in build_linux.sh. We'll need solve this if we want to make a build script on windows.

pavel-kirienko commented on test_linux.sh:
```diff
@@ -0,0 +1,11 @@
+timeout 20s ./zubax-kucher
+exit_status=$?
+if [ $exit_status -eq 124 ]; then
+    echo No error found
```
```suggestion
```

This doesn't really matter since this condition needs to be removed anyway, I just want to say that, typically, a well-behaved command-line tool reports nothing on success.

pavel-kirienko commented on test_linux.sh:
```diff
@@ -0,0 +1,11 @@
+
+timeout 20s ./zubax-kucher
+exit_status=$?
+if [ $exit_status -eq 124 ]; then
```
On GNU-based systems, any error code except zero indicates an error, so the test against 124 is not only unnecessary, it is actually dangerous because it is a case of fragile logic: the tool may return anything other than 124 to indicate an error, but this script would fail to catch that because it is hard coded to expect exactly 124.

Consider this approach instead: https://github.com/UAVCAN/pyuavcan/blob/d348a81fbfb96350305e18bf2188deecadde01d6/test.sh#L26-L49

loicgilbert replied:
My goal was to run the program for 20 seconds and if no error occured in kucher, it would return exit code 0. timeout return exit code 124 when the timer is over so it would make the build fail.

pavel-kirienko commented on test_windows.bat:
```diff
@@ -0,0 +1,13 @@
+start /B python zubax-kucher
+echo ok
```
Would help to be slightly more verbose here.

pavel-kirienko commented on test_linux.sh:
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+
+timeout 20s ./zubax-kucher
```
There is no point in just starting the program. The objective is to test its functionality. Run the unit tests here.

pavel-kirienko replied:
Come to think of it, just starting the program is not a useless test either. Keep it and run the unit tests, too.

pavel-kirienko commented:
There is a build script for Linux: https://github.com/Zubax/kucher/blob/master/build_linux.sh. It should be used in the CI to ensure that it's functional for local use.

Would it make sense to make a similar script for Windows?

loicgilbert committed:
Set up CI

pavel-kirienko commented on setup.cfg:
```diff
@@ -2,3 +2,36 @@
+ignore_missing_imports = True
+
+[mypy-numpy.*]
+ignore_missing_imports = True
```
This program does not use numpy, nunavut, or pydsdl; what's the point of listing them here?

pavel-kirienko committed:
Update README.md

pavel-kirienko commented:
@loicgilbert I noticed that the CI build is tied to your account on AppVeyor. Could you please create a Zubax org account and move the build there?

Also, we'll need a script with instructions detailing how to sign the Windows binary for redistribution. I'm going to leave the electronic signature dongle on your table.

pavel-kirienko committed:
Fixed the unit test invocation

pavel-kirienko committed:
Fixed some of the PEP8 violations

pavel-kirienko committed:
Merge pull request #25 from Zubax/pytest-fix

Fixed the unit tests

loicgilbert committed:
Updated Appveyor CI

loicgilbert committed:
Updated test files

loicgilbert committed:
PEP8 compatibility

loicgilbert committed:
Remove mypy

loicgilbert committed:
Add links to appveyor.yml

loicgilbert commented:
> Also, we'll need a script with instructions detailing how to sign the Windows binary for redistribution. I'm going to leave the electronic signature dongle on your table.

I have no idea how to do that. Any usefull link or ressource?

pavel-kirienko commented on appveyor.yml:
```diff
@@ -0,0 +1,53 @@
+  - cmd: "7z a zubax-kucher.7z *"
+  - sh: "7z a zubax-kucher.7z * &>/dev/null &"
+  - "appveyor PushArtifact zubax-kucher.7z"
+  - "pyinstaller pyinstaller.spec"
```
Invoking PyInstaller directly is not quite correct. On GNU/Linux, we must use StaticX to produce a truly portable executable, as implemented in this build script: https://github.com/Zubax/kucher/blob/appveyor-CI/build_linux.sh. Running just PyInstaller is not sufficient since the resulting executable ends up being [dependent on a particular version of libGL](https://github.com/Zubax/kucher/issues/1), which is disastrous for redistributable executables. So when building for GNU/Linux, invoke the above script. For consistency, I recommend defining an equivalent build script for Windows, despite the fact that no additional steps are needed there (yet).

Now, there was a problem with StaticX that I described here: https://github.com/JonathonReinhart/staticx/issues/79. Jonathon said that I should look into the dependencies to try and find out which one pulls in the unnecessary stuff. Shortly after receiving that reply I realized that the problem might be with the PyQtGraph library, which is no longer needed anyway (see https://github.com/Zubax/kucher/issues/21). I suggest to remove PyQtGraph and then see if StaticX is able to link a portable executable.

pavel-kirienko commented:
> I have no idea how to do that. Any usefull link or ressource?

@loicgilbert I'm going to pull this into a separate issue. No need to fix this in this PR.

pavel-kirienko commented:
To summarize, what's left here is:

- [x] comments about `QT_QPA_PLATFORM=offscreen`
- [x] build script for Windows
- [x] on GNU/Linux, use the build script instead of invoking PyInstaller directly due to portability issues

If the last one can't be fixed easily, we could comment out the staticx part temporarily and fix it in a separate pr later.

loicgilbert committed:
Add windows build file

loicgilbert committed:
Corrected appveyor.yml

loicgilbert committed:
Changed PyQt version

Install pyqt 5.12.2 instead of 5.12.3, causing errors when running executable (see github issue).

loicgilbert committed:
Fix staticx error

loicgilbert committed:
Fix  staticx

loicgilbert committed:
Added staticx

loicgilbert committed:
Exclude staticx from pep8

loicgilbert committed:
Fix 7z

pavel-kirienko commented:
Okay, almost done here. Having StaticX like this is not the best idea since it clutters our repo with a third-party codebase. What I meant when I said that we should fork it is that there should be a separate repository under the Zubax org which is a fork of StaticX; the fork can be created by clicking the aptly named button "Fork" in the upper-right side of the StaticX's GitHub page. GitHub will ask you where you want it forked, pick "Zubax" (I believe you have the necessary permissions for that). Then apply our patch there and add the patched repo as a submodule under `kucher/libraries`. That should be it.

pavel-kirienko commented:
Please also drop the PyQtGraph submodule from `kucher/libraries` as discussed, I just noticed that it's still there.

loicgilbert committed:
Merge pull request #24 from Zubax/pavel-kirienko-patch-1

Update README.md

loicgilbert committed:
Delete pyqtgraph

loicgilbert committed:
set staticx as a submodule

loicgilbert committed:
Update staticx

pavel-kirienko commented on build_linux.sh:
```diff
@@ -4,12 +4,20 @@
+pip install backports.lzma
 
 pyinstaller --clean --noconfirm pyinstaller.spec || exit 2
+cd kucher/libraries/staticx
```
```suggestion
pushd kucher/libraries/staticx
```

pavel-kirienko commented on build_linux.sh:
```diff
@@ -4,12 +4,20 @@
 
 # https://github.com/JonathonReinhart/staticx/issues/79
-cd dist
+cd ../../../dist
```
```suggestion
popd
pushd dist
```

pavel-kirienko commented on build_linux.sh:
```diff
@@ -4,12 +4,20 @@
 staticx --loglevel DEBUG Kucher.tmp Kucher
 
 rm -rf *.tmp
+cd ..
```
```suggestion
popd
```

pavel-kirienko commented on build_linux.sh:
```diff
@@ -4,12 +4,20 @@
 #
 
 sudo apt-get install patchelf -y
+sudo apt install gcc
```
```suggestion
sudo apt-get install -y gcc
```

Missing `-y` and changed apt to apt-get for consistency.

pavel-kirienko commented on build_linux.sh:
```diff
@@ -4,12 +4,20 @@
 
 sudo apt-get install patchelf -y
+sudo apt install gcc
+sudo apt-get install scons
```
```suggestion
sudo apt-get install -y scons
```

pavel-kirienko commented on .gitmodules:
```diff
@@ -13,3 +13,6 @@
 [submodule "kucher/libraries/quamash"]
```
The deleted pyqtgraph submodule needs to be removed from here also.

loicgilbert committed:
Update build_linux.sh and .gitmodule
